jwst/group_scale/group_scale_step.py

```diff
@@ -18,37 +18,40 @@ def process(self, input):
         # Open the input data model
         with datamodels.RampModel(input) as input_model:
 
+            # Always work on a copy
+            result = input_model.copy()
+
             # Try to get values of NFRAMES and FRMDIVSR to see
             # if we need to do any rescaling
-            nframes = input_model.meta.exposure.nframes
-            frame_divisor = input_model.meta.exposure.frame_divisor
+            nframes = result.meta.exposure.nframes
+            frame_divisor = result.meta.exposure.frame_divisor
 
             # If we didn't find NFRAMES, we don't have enough info
             # to continue. Skip the step.
             if nframes is None:
                 self.log.warning('NFRAMES value not found')
                 self.log.warning('Step will be skipped')
-                input_model.meta.cal_step.group_scale = 'SKIPPED'
-                return input_model
+                result.meta.cal_step.group_scale = 'SKIPPED'
+                return result
 
             # If we didn't find FRMDIVSR, then check to see if NFRAMES
             # is a power of 2. If it is, rescaling isn't needed.
             if frame_divisor is None:
                 if (nframes & (nframes - 1) == 0):
                     self.log.info('NFRAMES={} is a power of 2; correction not needed'.format(nframes))
                     self.log.info('Step will be skipped')
-                    input_model.meta.cal_step.group_scale = 'SKIPPED'
-                    return input_model
+                    result.meta.cal_step.group_scale = 'SKIPPED'
+                    return result
 
             # Compare NFRAMES and FRMDIVSR. If they're equal,
             # rescaling isn't needed.
             elif (nframes == frame_divisor):
                 self.log.info('NFRAMES and FRMDIVSR are equal; correction not needed')
                 self.log.info('Step will be skipped')
-                input_model.meta.cal_step.group_scale = 'SKIPPED'
-                return input_model
+                result.meta.cal_step.group_scale = 'SKIPPED'
+                return result
 
             # Do the scaling
-            result = group_scale.do_correction(input_model)
+            result = group_scale.do_correction(result)
```

Comment:
We make another copy inside of `group_scale.do_correction`:

https://github.com/spacetelescope/jwst/blob/master/jwst/group_scale/group_scale.py#L45

Should we skip that now?  Otherwise the result from this method will get overwritten by the result from `do_correction`, and the original result won't get closed.

Reply:
Agree to remove that copy; there is definitely enough copying going around.

However, a future discussion point: At this level, the code should not care whether a DataModel is closed or not: the user-level api should simply presume python scoping is in effect, end-of-story. If insufficient, that is actually a bug in the internals of DataModel, not the user code.

And FYI on this particular case: For the copied `result`, there would be nothing to close anyways. `DataModel.copy` is a deep copy (had to look it up), and no files are involved, so `DataModel.close` is a no-op, the in-memory stuff is still around, and won't be taken care of until gc-ed anyways.

Reply:
Oops, right!  Still can't keep it straight which one is deep copy and which is shallow 🤦

Reply:
Yeah, I think the choice of names could have been done wiser in the past.